docs/architecture/ADR-001-ABCI++.md

```diff
@@ -0,0 +1,414 @@
+
+While the adoption of ABCI++ is inevitable given the already made decision by upstream to implement it, here are some alternatives to the features that we need that do not use ABCI++.
+
+Alternatives for Message Inclusion
+https://github.com/celestiaorg/celestia-app/blob/92341dd68ee6e555ec6c0bb780afa3a1c8243a93/adrs/adr008:adopt-ABC%2B%2B-early.md#alternative-approaches
+
+Alternatives for Picking a square size
+https://github.com/celestiaorg/celestia-core/issues/454
+
+## Decision
+
+Proposed and initial implementation is complete.
+
+## Detailed Design
+
+### [#631](https://github.com/celestiaorg/celestia-core/pull/631) Simplified version of ABCI++ (`celestia-core`)
+
+Here we are adding only the two new methods that are necessary for the features that we need.  
+```go
+// Application is an interface that enables any finite, deterministic state machine
+// to be driven by a blockchain-based replication engine via the ABCI.
+type Application interface {
+   ...
+   PrepareProposal(RequestPrepareProposal) ResponsePrepareProposal
+   ProcessProposal(RequestProcessProposal) ResponseProcessProposal
+   ...
+}
+```
+
+It's also important to note the changes made to the request types for both methods. In upstream, they are only passing the transactions to the applications. This has been modifed to pass the entire block data.
+```protobuf
+message RequestPrepareProposal {
+ // block_data is an array of transactions that will be included in a block,
+ // sent to the app for possible modifications.
+ // applications can not exceed the size of the data passed to it.
+ tendermint.types.Data block_data = 1;
+ // If an application decides to populate block_data with extra information, they can not exceed this value.
+ int64 block_data_size = 2;
+}
+
+message RequestProcessProposal {
+ tendermint.types.Header header     = 1 [(gogoproto.nullable) = false];
+ tendermint.types.Data   block_data = 2;
+}
+
+...
+
+// Data contains the set of transactions, evidence, intermediate state roots,
+// and messages to be included in the block
+message Data {
+ // Txs that will be applied by state @ block.Height+1.
+ // NOTE: not all txs here are valid.  We're just agreeing on the order first.
+ // This means that block.AppHash does not include these txs.
+ repeated bytes txs = 1;
+
+ IntermediateStateRoots intermediate_state_roots = 2 [(gogoproto.nullable) = false];
+ EvidenceList           evidence                 = 3 [(gogoproto.nullable) = false];
+ Messages               messages                 = 4 [(gogoproto.nullable) = false];
+ uint64                 original_square_size     = 5;
+ bytes                  hash                     = 6; // <-- The hash has been added so that the erasure step can be performed in the app
+}
+```
+
+Here is where the new ABCI++ method is called to perform arbitrary logic in the app before creating the proposal block. `consensus/state.go`
+
+```go
+// CreateProposalBlock calls state.MakeBlock with evidence from the evpool
```

Comment:
what is `evpool` ?

Reply:
evidence pool https://github.com/celestiaorg/celestia-core/blob/86b292b620b34a360fe3fc754ec55d0b42bec400/internal/evidence/pool.go#L29-L52